### ISSUE TITLE
domain / project memories

### ISSUE THREAD

**Issue by pdhoolia** [pdhoolia/se-agent#11]
Lack of domain understanding may sometime lead to suggestions that are not feasible or practical. For example, suggesting a change in an emulator (that is supposed to exactly replicate quantum hardware) may not be the right suggestion to fix a bug in qiskit. Or something that is not possible in the current version of the programming language, or suggesting a change that is not in line with the project's coding standards.

When agent makes such proposals, other developers may point out the issues with the suggestion via comments. The agent should be able to understand these comments and their context and learn from them. It should be able to create a succinct, condensed memory of this understanding and store it for future reference.

Here's a high-level plan to implement this feature:

- Before starting to process an issue comment, the agent may analyze the comment to check if it represents a domain understanding correction. This may be done by checking if the comment may be semantically categorized as a domain characteristic. Note: The agent probably uses the term `project` (for domain).

- If comment is identified as a domain understanding correction, the agent should add new functionality to generate a condensed memory of this understanding.
    - It should use the issue conversation context to generate this memory.
    - It should add a new LLM task for this purpose, form a prompt, and call the LLM to generate memories (there could be more than 1). Each memory should be condensed (e.g., no more than 20 tokens)
    - Agent should store these memories in a memory store. For the first version, we can use a json file in the metadata folder for storing memories.
    - Each memory should be stored as a json object with the following fields:
        - `memory_id`: An id (generated) for the memory
        - `issue_id`: The issue id in the context of which this memory was generated
        - `comment_id`: The comment id based on which this memory was generated
        - `memory`: The memory text
        - `user_id`: The user id who made the comment
        - `created_at`: The timestamp when this memory was created
        - `updated_at`: The timestamp when this memory was last updated
        - `status`: At the time of creation, the status of the memory should be `new`. This status may be updated later based on curation by an expert. Expert may mark it as `accepted`, `rejected`, `needs_review` etc.

- The agent should retrieve these memories when processing new issues or comments. In particular, it should use them in the prompts for localization and suggestions.

**Comment by se-agent** [comment]
To implement this feature for creating and utilizing domain/project memories, we will need to modify existing components and possibly introduce new modules. Here's a step-by-step approach to achieve the desired functionality:

1. **Identify Domain Understanding Corrections:**
   - Modify the `process_issue_event` function in `listener_core.py` to analyze comments for domain understanding corrections.
   - Introduce a new function `is_domain_correction_comment` to determine if a comment addresses a domain understanding correction. This function will use semantic analysis, possibly leveraging an LLM task.

2. **Generate and Store Domain Memories:**
   - Create a new LLM task in `llm/api.py` to generate condensed domain memories based on issue conversations. This task will return succinct memories (e.g., not more than 20 tokens each).
   - Add a function `generate_domain_memories` within `listener_core.py` that:
     - Forms a prompt using the issue's conversation context.
     - Calls the new LLM task to generate memories.
     - Stores these memories in a JSON file within a designated metadata folder. This will involve creating a new file `memory_manager.py` to handle memory storage and retrieval operations.

3. **Store Memories as JSON Objects:**
   - Define a data model for storing memories, which includes fields like `memory_id`, `issue_id`, `comment_id`, `memory`, `user_id`, `created_at`, `updated_at`, and `status`.
   - Implement functions to create, update, and retrieve memories in `memory_manager.py`.

4. **Retrieve and Use Memories:**
   - Update `localize_issue` and `suggest_changes` functions to retrieve relevant domain memories from the memory store.
   - Incorporate these memories in prompts sent to the LLMs to aid in localization and suggestion generation.

Here's a concise code change plan:

### Code Changes

1. **listener_core.py**
   - Add a check for domain corrections in `process_issue_event`.
   - Call `generate_domain_memories` if a domain correction is identified.

2. **llm/api.py**
   - Implement a new LLM task, `GENERATE_MEMORIES`, for creating domain memories.
   - Add logic to handle this new task in `call_llm_for_task`.

3. **memory_manager.py**
   - Define data models for memories.
   - Implement functions to create, update, and retrieve memories.

4. **localizer.py** and **change_suggester.py**
   - Update functions to incorporate domain memories in LLM prompts for better localization and suggestion accuracy.

### Sample Code Snippets

**listener_core.py**
```python
def is_domain_correction_comment(comment_body):
    # This function will use semantic analysis to determine if the comment is a domain correction
    # For now, we can implement a placeholder that checks for specific keywords or patterns
    domain_keywords = ["not feasible", "incorrect", "standard", "impossible"]
    return any(keyword in comment_body.lower() for keyword in domain_keywords)

def generate_domain_memories(project, issue_details, comment_details):
    # Form a prompt using the conversation context
    conversation = issue_analyzer.get_conversation_context(issue_details, comment_details)
    prompt = create_memory_prompt(conversation)
    
    # Call LLM to generate domain memories
    memories = call_llm_for_task(TaskName.GENERATE_MEMORIES, prompt)
    
    # Store memories using memory manager
    memory_manager.store_memories(issue_details['number'], comment_details['id'], memories)

# Inside process_issue_event
if comment_details and is_domain_correction_comment(comment_details['body']):
    generate_domain_memories(project, issue_details, comment_details)
```

**memory_manager.py**
```python
import json
import os
from datetime import datetime

MEMORY_FILE_PATH = os.path.join('metadata', 'domain_memories.json')

def store_memories(issue_id, comment_id, memories):
    current_time = datetime.now().isoformat()
    memory_entries = []
    for memory in memories:
        memory_entry = {
            "memory_id": generate_memory_id(),
            "issue_id": issue_id,
            "comment_id": comment_id,
            "memory": memory,
            "user_id": "unknown",  # Retrieve user ID if available
            "created_at": current_time,
            "updated_at": current_time,
            "status": "new"
        }
        memory_entries.append(memory_entry)
    append_to_memory_file(memory_entries)

def append_to_memory_file(memory_entries):
    if not os.path.exists(MEMORY_FILE_PATH):
        with open(MEMORY_FILE_PATH, 'w') as f:
            json.dump(memory_entries, f, indent=4)
    else:
        with open(MEMORY_FILE_PATH, 'r+') as f:
            existing_data = json.load(f)
            existing_data.extend(memory_entries)
            f.seek(0)
            json.dump(existing_data, f, indent=4)

def generate_memory_id():
    # Generate a unique ID for the memory
    return str(uuid.uuid4())
```

These changes outline the basic structure required to implement domain/project memories. You will need to expand on these based on the specifics of your project and LLM capabilities.

**Comment by pdhoolia** [comment]
can you format the `listener_core.py` change in git-diff format?

**Comment by se-agent** [comment]
Certainly! Below is the `git diff` formatted change for `listener_core.py`, incorporating the new functionality for detecting and generating domain memories:

```diff
diff --git a/se_agent/listener_core.py b/se_agent/listener_core.py
index abcdef1..1234567 100644
--- a/se_agent/listener_core.py
+++ b/se_agent/listener_core.py
@@ -1,6 +1,7 @@
 import json
 import os
 import logging
+from se_agent.memory_manager import store_memories
 logger = logging.getLogger("se-agent")

 from se_agent.issue_analyzer import analyze_issue
@@ -93,6 +94,7 @@ def process_issue_event(project: Project, issue_details, comment_details=None):
         
         # Ignore comments that are from the agent itself
         if '<!-- SE Agent -->' in comment_body:
+            print("Ignoring agent's own comment")
             return IGNORE_TOKEN
         
         # Ignore comments that close the issue
@@ -111,6 +113,9 @@ def process_issue_event(project: Project, issue_details, comment_details=None):
 
     # Analyze, localize, and suggest changes for the issue
     try:
+        # Check for domain understanding corrections
+        if comment_details and is_domain_correction_comment(comment_body):
+            generate_domain_memories(project, issue_details, comment_details)
         analysis_results = analyze_issue(project, issue_details)
         logger.debug(f"Analysis results: {analysis_results}")
         localization_results = localize_issue(project, issue_details, analysis_results)
@@ -147,3 +152,28 @@ def process_issue_event(project: Project, issue_details, comment_details=None):
         project.post_issue_comment(issue_number, agent_comment)
     except Exception as e:
         logger.exception("Error posting comment to GitHub.")
+        return None
+
+    return change_suggestions
+
+def is_domain_correction_comment(comment_body):
+    """
+    Determines if a comment is a domain understanding correction.
+    This function uses semantic analysis to find specific keywords or patterns.
+    """
+    domain_keywords = ["not feasible", "incorrect", "standard", "impossible"]
+    return any(keyword in comment_body.lower() for keyword in domain_keywords)
+
+def generate_domain_memories(project: Project, issue_details, comment_details):
+    """
+    Generates and stores domain memories based on issue conversation context.
+    """
+    # Form a prompt using the conversation context
+    conversation = analyze_issue.get_conversation_context(issue_details, comment_details)
+    prompt = create_memory_prompt(conversation)
+
+    # Call LLM to generate domain memories
+    memories = call_llm_for_task(TaskName.GENERATE_MEMORIES, prompt)
+
+    # Store memories using memory manager
+    store_memories(issue_details['number'], comment_details['id'], memories)
```

This `git diff` output shows the changes made to `listener_core.py`, including the addition of the `is_domain_correction_comment` and `generate_domain_memories` functions, and the integration of these functions into the existing code flow.
